src/base/bittorrent/nativesessionextension.cpp

```diff
@@ -30,26 +30,19 @@
 
 #include <libtorrent/alert_types.hpp>
 
+#include "base/bittorrent/torrentinfo.h"
 #include "nativetorrentextension.h"
 
-namespace
-{
-    void handleFastresumeRejectedAlert(const lt::fastresume_rejected_alert *alert)
-    {
-        if (alert->error.value() == lt::errors::mismatching_file_size) {
-            alert->handle.unset_flags(lt::torrent_flags::auto_managed);
-            alert->handle.pause();
-        }
-    }
-}
-
 lt::feature_flags_t NativeSessionExtension::implemented_features()
 {
     return alert_feature;
 }
 
-std::shared_ptr<lt::torrent_plugin> NativeSessionExtension::new_torrent(const lt::torrent_handle &torrentHandle, ClientData)
+std::shared_ptr<lt::torrent_plugin> NativeSessionExtension::new_torrent(const lt::torrent_handle &torrentHandle, ClientDataPtr clientDataPtr)
 {
+    if (clientDataPtr)
```

Comment:
I'm compiling with libtorrent 2.0 (with deprecated functions ON) and this line gives an error:
```
In file included from /usr/include/libtorrent/torrent_handle.hpp:70,
                 from /usr/include/libtorrent/torrent_status.hpp:40,
                 from /usr/include/libtorrent/extensions.hpp:42,
                 from ../../src/base/bittorrent/nativesessionextension.h:31,
                 from ../../src/base/bittorrent/nativesessionextension.cpp:29:
/usr/include/libtorrent/client_data.hpp: In instantiation of ‘libtorrent::client_data_t::operator T() const [with T = bool]’:
../../src/base/bittorrent/nativesessionextension.cpp:43:22:   required from here
/usr/include/libtorrent/client_data.hpp:77:37: error: static assertion failed: client_data_t can only be cast to a pointer
   77 |   static_assert(std::is_pointer<T>::value, "client_data_t can only be cast to a pointer");
      |                                     ^~~~~
/usr/include/libtorrent/client_data.hpp:78:75: error: converting to ‘bool’ from ‘std::nullptr_t’ requires direct-initialization [-fpermissive]
   78 |   if (m_type_ptr != type<typename std::remove_pointer<T>::type>()) return nullptr;
      |                                                                           ^~~~~~~
```

In libtorrent 2.0, seems you should write
```c++
if (clientDataPtr.get<void *>())
// or
if (static_cast<void *>(clientDataPtr))
```

Reply:
Off topic:
Since we are now trying to be compatible with libtorrent 2.0, I plan to add another libtorrent branch in TravisCI script (with deprecated functions ON at this moment). What do you think?

Reply:
I agree. But I wouldn't add its counterpart to every existing configuration. It's already taking too long to process.

Reply:
This is a use case I hadn't considered. The reason for this failing is because this will invoke the template conversion operator to bool, since the if-condition creates a boolean context.

That conversion operator should not have a static assert in it, it should just be SFINAEd out for non-pointer Ts. And maybe there should be an explicit bool conversion operator too, for this kind of use.

Reply:
also, to test if `ClientDataPtr` is set, you specify the type you expect it to have. If it set to a different type, you'll still gett `nullptr` back. So the fix for that if-statement is:

```
if (ClientDataPtr.get<ClientDataImpl *>())
```

I'm adding a bool conversion operator for the next 2.0 release too

Reply:
actually, I'm not sure about a bool conversion operator, it would be quite misleading if it's later cast to a different type yielding `nullptr`. I think, in this code, to be compatible both with `client_data_t` and `void*`, casting to a new variable of the expected pointer type, and then test that and use that variable would be the safest.

Reply:
I wonder why all this trouble with `client_data_t`? Is it supposed that different torrents within the same session will use different types of data?

Reply:
that's a possibility, but probably not very likely. But certainly different clients of libtorrent will use different types. This provides type-safety

Reply:
Clear. Just another new-fangled thing. I don't see any problems if client uses a single type.